internal/controller/factory/statefulset.go

```diff
@@ -70,6 +71,31 @@ func CreateOrUpdateStatefulSet(
 
 	volumes := generateVolumes(cluster)
 
+	basePodSpec := corev1.PodSpec{
+		Containers: []corev1.Container{generateContainer(cluster)},
+		Volumes:    volumes,
+	}
+	basePodSpecBytes, err := json.Marshal(basePodSpec)
+	if err != nil {
+		return fmt.Errorf("cannot marshal generated base pod spec to JSON: %w", err)
+	}
+	if cluster.Spec.PodTemplate.Spec.Containers == nil {
+		cluster.Spec.PodTemplate.Spec.Containers = make([]corev1.Container, 0)
+	}
```

Comment:
It so happens that the JSON marshal of an empty PodSpec is not `{}`. but `{"containers":null}` (the only field without omitEmpty, I guess). So I have to explicitly set it to an empty list, or else `StrategicMergePatch` deletes the `containers` field in the final statefulset.